src/core/ext/filters/client_channel/subchannel_index.cc

```diff
@@ -49,6 +49,8 @@ static grpc_subchannel_key* create_key(
     grpc_channel_args* (*copy_channel_args)(const grpc_channel_args* args)) {
   grpc_subchannel_key* k =
       static_cast<grpc_subchannel_key*>(gpr_malloc(sizeof(*k)));
+  // Currently, args->filter_count is always 0, which means we only use the
```

Comment:
Good catch!  This looks like a bug to me -- it would actually be bad if we considered two subchannels with different filter lists to be the same.

We've probably gotten away with this because in general, it's the set of channel args that determines the set of filters.  However, I think there are some cases (especially internally) where a filter registration function will decide whether or not to add a filter based on something other than channel args.  But you can verify this by looking at all existing filter registration functions (i.e., functions passed to `grpc_channel_init_register_stage()` to register filters, or C++ filters using `grpc::RegisterChannelFilter()`).

To fix this, I think we should do one of two things:

1. Fix the code somehow to actually populate `args->filter_count` and `args->filters` before creating the subchannel key, so that we ensure that we don't treat subchannels with different filters as the same.

2. Decide that we want to require that subchannels with different filter lists will always have different channel args.  In this case, we should document this restriction for registration functions.  (Note that even if a registration function uses some external mechanism to actually make the decision, it could then *set* a different channel arg to reflect its decision, so the channel args would always be different if the set of filters are different.)  We should also remove the `filters` and `filter_count` fields from `grpc_subchannel_args`.  (Actually, at that point, the only field left in `grpc_subchannel_args` would be the channel args, so we could just replace `grpc_subchannel_args` with `grpc_channel_args`.)

Reply:
This is hard to achieve, because the subchannel stack is created after it's connected.

When we ~create~ a subchannel, we use the channel args (basically the same one as the `chand->resolver_result`) to create the subchannel key. And we don't know any thing about the filters yet.

When the subchannel is ~connected~, we will create the channel stack in https://github.com/grpc/grpc/blob/6a2fbcb4d27243e8c32227b681c38c6221e09a00/src/core/ext/filters/client_channel/subchannel.cc#L586. The channel stack builder will create the filters based on `c->connecting_result.channel_args` and `c->connecting_result.transport`. 

So, when we create the subchannel key, we don't know what filters a subchannel will have.

Also, that why I feel it's hard to pre-compute the size of the subchannel stack and add it into the call size estimate.

Reply:
I agree that option 1 is challenging; see my comments in #16248.  But how about option 2?